File: 200-number-of-islands/number-of-islands.py
# traverse the grid
# if the current cell is a 1
# incrment the count of island and explore the entire island

class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:

        count = 0
        rows = len(grid)
        columns = len(grid[0])

        def traverse(i, j):
            if (i < 0 or i == rows or j < 0 or j == columns or grid[i][j] != "1"):
                return 

            grid[i][j] = "0"
            traverse(i + 1, j)
            traverse(i - 1, j)
            traverse(i, j + 1)
            traverse(i, j - 1)
        

        for i in range(rows):
            for j in range(columns):
                cell = grid[i][j]

                if (cell == "1"):
                    count += 1
                    traverse(i, j)

        
        return count
        

File: 200-number-of-islands/number-of-islands.js
// run dfs on the grid
// count the number of islands

var numIslands = function(grid) {
    let rows = grid.length
    let columns = grid[0].length
    directions = [[-1, 0], [0, 1], [1, 0], [0, -1]]
    count = 0

    function isValid(i, j) {
        if (i < 0 || i == rows || j < 0 || j == columns || grid[i][j] != "1") return false
        grid[i][j] = "0"
        return true
    }

    function traverse(i, j) {
        for (let direction of directions) {
            let [x, y] = direction
            if (isValid(i + x, y + j)) traverse(i + x, y + j)
        }
    }   

    for (let i = 0; i < rows; i++) {
        for (let j = 0; j < columns; j++) {
            if (grid[i][j] == "1") {
                count +=1
                traverse(i, j)
            }
        }
    }

    return count






};